GitVersionCore/VersionCalculation/NewNextVersionCalculator.cs

```diff
@@ -0,0 +1,74 @@
+﻿namespace GitVersion.VersionCalculation
+{
+    using System;
+    using System.Text.RegularExpressions;
+    using BaseVersionCalculators;
+
+    public class NewNextVersionCalculator
+    {
+        IBaseVersionCalculator baseVersionFinder;
+        IMetaDataCalculator metaDataCalculator;
+
+        public NewNextVersionCalculator(IBaseVersionCalculator baseVersionCalculator = null, IMetaDataCalculator metaDataCalculator = null)
+        {
+            this.metaDataCalculator = metaDataCalculator ?? new MetaDataCalculator();
+            baseVersionFinder = baseVersionCalculator ??
+                new BaseVersionCalculator(
+                    new FallbackBaseVersionStrategy(),
+                new ConfigNextVersionBaseVersionStrategy(),
+                new LastTagBaseVersionStrategy(),
+                new MergeMessageBaseVersionStrategy(),
+                new VersionInBranchBaseVersionStrategy());
+        }
+
+        public SemanticVersion FindVersion(GitVersionContext context)
+        {
+            var baseVersion = baseVersionFinder.GetBaseVersion(context);
+
+            if (baseVersion.ShouldIncrement) IncrementVersion(context, baseVersion);
+
+            if (baseVersion.ShouldUpdateTag && !baseVersion.SemanticVersion.PreReleaseTag.HasTag() && !string.IsNullOrEmpty(context.Configuration.Tag))
+            {
+                var tagToUse = context.Configuration.Tag;
+                if (tagToUse == "useBranchName")
```

Comment:
I thought about that, but it is a lot harder.

If we specify defaults, then you override do we clear out the `tag` option. If the user sets `useBranchNameAsTag` to true and also sets `tag` do we throw, or ignore the tag?